### PR TITLE
feat: follow button in the list

### LINKED ISSUE #385: 在 Follow List 加入快捷关注按钮
有时候想通过别人的 Follow List 来拓展自己的好友圈，希望在这一排加入关注按钮，直接点击关注，更方便

image

### PR DESCRIPTION
**Pull request by suemor233**
### <samp>🤖 Generated by Copilot at 104501e</samp>

The pull request enhances the `CharacterList` component to enable following and unfollowing characters from different sites, and updates the `Modal` component to use the clsx library and a ZIndex prop. These changes improve the user experience and the code quality of the xLog app.

### <samp>🤖 Generated by Copilot at 104501e</samp>

> _Oh we're the crew of the CharacterList_
> _We follow and unfollow as we wish_
> _We heave and we ho with the clsx_
> _And we set the ZIndex of the Modal_

### WHY

close #385 

### <samp>🤖 Generated by Copilot at 104501e</samp>

* Add a ZIndex prop to the Modal component to control its stacking order ([link](https://github.com/Crossbell-Box/xLog/pull/388/files?diff=unified&w=0#diff-a409ba5ed93668907a24feb7c7ebfd9866fde92d6bd14a811c66638e19594c09L27-R31), [link](https://github.com/Crossbell-Box/xLog/pull/388/files?diff=unified&w=0#diff-0c04326a03a8a8e00e64e35c3edd62da56cb506f702f2a70b98b525e880eac98L14-R21))
* Import the clsx library in `Modal.tsx` to apply conditional class names based on the ZIndex prop ([link](https://github.com/Crossbell-Box/xLog/pull/388/files?diff=unified&w=0#diff-0c04326a03a8a8e00e64e35c3edd62da56cb506f702f2a70b98b525e880eac98R1))
* Import React, Profile, siteModel, and FollowingButton in `CharacterList.tsx` to use them in the CharacterList component ([link](https://github.com/Crossbell-Box/xLog/pull/388/files?diff=unified&w=0#diff-a409ba5ed93668907a24feb7c7ebfd9866fde92d6bd14a811c66638e19594c09R2), [link](https://github.com/Crossbell-Box/xLog/pull/388/files?diff=unified&w=0#diff-a409ba5ed93668907a24feb7c7ebfd9866fde92d6bd14a811c66638e19594c09L9-R16))
* Modify the layout and style of the character items in the list, adding a FollowingButton component and adjusting the overflow and spacing of the text elements ([link](https://github.com/Crossbell-Box/xLog/pull/388/files?diff=unified&w=0#diff-a409ba5ed93668907a24feb7c7ebfd9866fde92d6bd14a811c66638e19594c09L45-R86))
* Define a FollowingButtonItem component that fetches the site data and renders a FollowingButton component for each character item ([link](https://github.com/Crossbell-Box/xLog/pull/388/files?diff=unified&w=0#diff-a409ba5ed93668907a24feb7c7ebfd9866fde92d6bd14a811c66638e19594c09R106-R116))

### CODE REVIEW

**Comment by suemor233** [comment]
image

File: src/components/common/CharacterListItem.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import * as siteModel from "~/models/site.model"
+
+import { Avatar } from "../ui/Avatar"
+import { UniLink } from "../ui/UniLink"
+import { FollowingButton } from "./FollowingButton"
+
+export const CharacterListItem: React.FC<{
+  character: any
+  sub: any
+}> = ({ character, sub }) => {
```
Comment (Innei) [inline src/components/common/CharacterListItem.tsx:17]: React 编写范式：此组件用在 循环 中，外层包一个 `memo`

File: src/components/common/CharacterListItem.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  const [site, setSite] = useState<Profile>()
+  const siteId = character?.handle
+
+  useEffect(() => {
+    if (siteId) {
+      siteModel.getSite(siteId).then((site) => setSite(site))
```
Comment (Innei) [inline src/components/common/CharacterListItem.tsx:23]: 可以用 react-query?

Comment (Innei) [reply src/components/common/CharacterListItem.tsx:23]: @DIYgod 没有可以聚合的接口获取状态么，每个都会打一个接口，量多了会有性能问题。

**Review by Innei — CHANGES_REQUESTED** [review]

**Comment by Innei** [comment]
Next.js Bundle Analysis / analyze (pull_request 这个 CI 为啥一直都是挂的。 

> Resource not accessible by integration

好奇怪。

**Comment by Innei** [comment]
> Next.js Bundle Analysis / analyze (pull_request 这个 CI 为啥一直都是挂的。
> 
> > Resource not accessible by integration
> 
> 好奇怪。

Bot 的 pr 都正常。。

**Comment by DIYgod** [comment]
You can use GraphQL to reduce the number of requests. Here is the [documentation](https://indexer.crossbell.io/altair), and below is example.

```GraphQL
query getFollows {
  links(
    where: {
      linkType: { equals: "follow" },
      fromCharacterId: { equals: 10 },
      toCharacterId: { in: [52343, 54430] }
    },
  ) {
    toCharacterId
  }
}
```

File: src/components/common/CharacterListItem.tsx (hunk 2 above)
Comment (DIYgod) [reply src/components/common/CharacterListItem.tsx]: 有的，我在下面评论啦

**Comment by suemor233** [comment]
> You can use GraphQL to reduce the number of requests. Here is the [documentation](https://indexer.crossbell.io/altair), and below is example.
> 
> ```graphql
> query getFollows {
>   links(
>     where: {
>       linkType: { equals: "follow" },
>       fromCharacterId: { equals: 10 },
>       toCharacterId: { in: [52343, 54430] }
>     },
>   ) {
>     toCharacterId
>   }
> }
> ```

I found that the `<CharacterListItem/>` props already have the props needed for the `<FollowingButton/>`, so I don't need to send any requests to complete it.

File: src/components/ui/Modal.tsx [hunk 3]
```diff
@@ -11,12 +12,13 @@ export const Modal: React.FC<{
   title?: string | React.ReactNode
   titleIcon?: React.ReactNode
   size?: "md" | "lg" | "sm"
-}> = ({ open, setOpen, children, title, titleIcon, size = "md" }) => {
+  ZIndex?: number
```
Comment (Innei) [inline src/components/ui/Modal.tsx:15]:
```suggestion
 zIndex?: number
```

**Review by Innei — CHANGES_REQUESTED** [review]

**Comment by DIYgod** [comment]
Thanks so much!

**Pull request merged**